File: src/test/scala/LeagueRankerTest.scala
package leaderboard

import org.scalatest.funsuite.AnyFunSuite
import leaderboard.TeamResult

class LeagueRankerTest extends AnyFunSuite {

  test("splitTeamResult should parse valid input") {
    val result = LeagueRanker.parseTeamResult("Lions 3")
    assert(result.isSuccess)
    assert(result.get == TeamResult("Lions", 3))
  }

  test("splitTeamResult should fail on invalid input") {
    val result = LeagueRanker.parseTeamResult("Broken")
    assert(result.isFailure)
  }

  test("validateLine should parse valid line") {
    val line = "Lions 2, Snakes 1"
    val result = LeagueRanker.validateLine(line)
    assert(result.isSuccess)
    val (team1, team2) = result.get
    assert(team1 == TeamResult("Lions", 2))
    assert(team2 == TeamResult("Snakes", 1))
  }

  test("validateLine should fail on invalid line") {
    val line = "InvalidLine"
    val result = LeagueRanker.validateLine(line)
    assert(result.isFailure)
  }

  test("rankTeams returns correctly ranked output") {
    val scores = Map(
      "Tarantulas" -> 6,
      "Lions" -> 5,
      "FC Awesome" -> 1,
      "Snakes" -> 1,
      "Grouches" -> 0
    )
    val expected = List(
      "1. Tarantulas, 6 pts",
      "2. Lions, 5 pts",
      "3. FC Awesome, 1 pt",
      "3. Snakes, 1 pt",
      "5. Grouches, 0 pts"
    )
    val actual = LeagueRanker.rankTeams(scores)
    assert(actual == expected)
  }
}

File: src/test/java/com/leaderboard/LeagueRankerTest.java
package com.leaderboard;

import org.junit.jupiter.api.Test;

import java.util.*;

import static org.junit.jupiter.api.Assertions.*;

class LeagueRankerTest {

    @Test
    void testParseTeamValid() {
        var result = LeagueRanker.parseTeam("Lions 3");
        assertTrue(result.isPresent());
        assertEquals("Lions", result.get().name());
        assertEquals(3, result.get().score());
    }

    @Test
    void testParseTeamInvalid() {
        assertTrue(LeagueRanker.parseTeam("InvalidLine").isEmpty());
    }

    @Test
    void testComputeScoresAndRanking() {
        List<String> input = List.of(
                "Lions 3, Snakes 3",
                "Tarantulas 1, FC Awesome 0",
                "Lions 1, FC Awesome 1",
                "Tarantulas 3, Snakes 1",
                "Lions 4, Grouches 0"
        );
        List<String> errors = new ArrayList<>();
        Map<String, Integer> scores = LeagueRanker.computeScores(input, errors);

        List<String> expected = List.of(
                "1. Tarantulas, 6 pts",
                "2. Lions, 5 pts",
                "3. FC Awesome, 1 pt",
                "3. Snakes, 1 pt",
                "5. Grouches, 0 pts"
        );

        List<String> actual = LeagueRanker.rankTeams(scores);
        assertEquals(expected, actual);
    }
}
